src/drivers/legos/driver.rs

```diff
@@ -0,0 +1,404 @@
+        let mut buf = [0; XINPUT_PACKET_SIZE];
+        let bytes_read = self.device.read_timeout(&mut buf[..], HID_TIMEOUT)?;
+
+        let mut events;
+
+        if bytes_read != XINPUT_PACKET_SIZE {
+            // controller device goes to sleep and reads time out after a while
+            // todo: check actual timeout?, but either way shouldn't be fatal
+            events = vec![];
+        } else {
+            events = self.handle_xinput_report(buf)?;
+        }
+
+        // There is no release event, so check to see if we are still touching.
+        if self.is_touching && (self.last_touch.elapsed() > Duration::from_millis(4)) {
+            let event: Event = self.release_touch();
+            events.push(event);
+            // Check for tap events
+            if self.first_touch.elapsed() < Duration::from_millis(200) {
+                // For double clicking, ensure the previous tap is cleared.
+                if self.is_tapped {
+                    let event: Event = self.release_tap();
+                    events.push(event);
+                }
+                let event: Event = self.start_tap();
+                events.push(event);
+            }
+        }
+
+        // If we did a click event, see if we shoudl release it. Accounts for click and drag.
+        if !self.is_touching
+            && self.is_tapped
+            && (self.last_touch.elapsed() > Duration::from_millis(100))
+        {
+            let event: Event = self.release_tap();
+            events.push(event);
+        }
+
+        Ok(events)
+    }
+
+    /// Unpacks the buffer into a [XinputDataReport] structure and updates
+    /// the internal xinput_state
+    fn handle_xinput_report(
+        &mut self,
+        buf: [u8; XINPUT_PACKET_SIZE],
+    ) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        let input_report = XInputDataReport::unpack(&buf)?;
+
+        // Print input report for debugging
+        //log::trace!("--- Input report ---");
+        //log::trace!("{input_report}");
+        //log::trace!(" ---- End Report ----");
+
+        // Update the state
+        let old_dinput_state = self.update_xinput_state(input_report);
+
+        // Translate the state into a stream of input events
+        let events = self.translate_xinput(old_dinput_state);
+
+        Ok(events)
+    }
+
+    /// Update gamepad state
+    fn update_xinput_state(&mut self, input_report: XInputDataReport) -> Option<XInputDataReport> {
+        let old_state = self.xinput_state;
+        self.xinput_state = Some(input_report);
+        old_state
+    }
+
+    /// Translate the state into individual events
+    fn translate_xinput(&self, old_state: Option<XInputDataReport>) -> Vec<Event> {
+        let mut events = Vec::new();
+        let Some(state) = self.xinput_state else {
+            return events;
+        };
+
+        // Translate state changes into events if they have changed
+        if let Some(old_state) = old_state {
```

Comment:
Nit: we can reduce nesting here using this pattern instead of `if let Some(old_state)`:

```rust
let Some(old_state) = old_state else {
    return events;
};

// Binary Events
if state.a != old_state.a {
    events.push(Event::GamepadButton(GamepadButtonEvent::A(BinaryInput {
        pressed: state.a,
    })));
}
...
```